doc/ssh_filter_btrbk.1
'\" t
.\"     Title: ssh_filter_btrbk
.\"    Author: [see the "AUTHOR(S)" section]
.\" Generator: Asciidoctor 1.5.7.1
.\"      Date: 2019-04-24
.\"    Manual: Btrbk Manual
.\"    Source: Btrbk 0.28.0
.\"  Language: English
.\"
.TH "SSH_FILTER_BTRBK" "1" "2019-04-24" "Btrbk 0.28.0" "Btrbk Manual"
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.ss \n[.ss] 0
.nh
.ad l
.de URL
\fI\\$2\fP <\\$1>\\$3
..
.als MTO URL
.if \n[.g] \{\
.  mso www.tmac
.  am URL
.    ad l
.  .
.  am MTO
.    ad l
.  .
.  LINKSTYLE blue R < >
.\}
.SH "NAME"
ssh_filter_btrbk \- ssh command filter script for btrbk
.SH "SYNOPSIS"
.sp
.nf
ssh_filter_btrbk.sh [\-s|\-\-source] [\-t|\-\-target] [\-d|\-\-delete]
    [\-i|\-\-info] [\-\-snapshot] [\-\-send] [\-\-receive]
    [\-p|\-\-restrict\-path <path>] [\-l|\-\-log] [\-\-sudo]
.fi
.br
.SH "DESCRIPTION"
.sp
\fBssh_filter_btrbk.sh\fP restricts SSH commands to commands used by
\fIbtrbk\fP. It examines the SSH_ORIGINAL_COMMAND environment variable
(set by sshd) and executes it only if it contains commands used by
\fIbtrbk\fP.
.sp
The accepted commands are specified by the "\-\-source", "\-\-target",
"\-\-delete" and "\-\-info" options.
.sp
The following commands are always allowed:
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.  sp -1
.  IP \(bu 2.3
.\}
"btrfs subvolume show"
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.  sp -1
.  IP \(bu 2.3
.\}
"btrfs subvolume list"
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.  sp -1
.  IP \(bu 2.3
.\}
"readlink"
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.  sp -1
.  IP \(bu 2.3
.\}
"cat /proc/self/mountinfo"
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.  sp -1
.  IP \(bu 2.3
.\}
pipes through "gzip", "pigz", "bzip2", "pbzip2", "xz", "lzop",
"lz4" (stream_compress)
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.  sp -1
.  IP \(bu 2.3
.\}
pipes through "mbuffer" (stream_buffer)
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.  sp -1
.  IP \(bu 2.3
.\}
pipes through "pv \-L" (rate_limit)
.RE
.sp
Example line in /root/.ssh/authorized_keys on a backup target host:
.sp
.if n .RS 4
.nf
command="ssh_filter_btrbk.sh \-\-target \-\-delete \-\-restrict\-path /mnt/btr_backup" ssh\-rsa AAAAB3NzaC1...hwumXFRQBL btrbk@mydomain.com
.fi
.if n .RE
.SH "OPTIONS"
.sp
\-s, \-\-source
.RS 4
Allow commands for backup source: "btrfs subvolume snapshot",
"btrfs send". Equivalent to "\-\-snapshot \-\-send".
.RE
.sp
\-t, \-\-target
.RS 4
Allow commands for backup target: "btrfs receive".
.RE
.sp
\-d, \-\-delete
.RS 4
Allow commands for subvolume deletion: "btrfs subvolume
delete". This is used for backup source if
\fIsnapshot_preserve_daily\fP is not set to \(lqall\(rq, and for backup
targets if \fItarget_preserve_daily\fP is not set to \(lqall\(rq.
.RE
.sp
\-i, \-\-info
.RS 4
Allow informative commands: "btrfs subvolume find\-new", "btrfs
filesystem usage". This is used by btrbk \fIinfo\fP and \fIdiff\fP
commands.
.RE
.sp
\-\-snapshot
.RS 4
Allow btrfs snapshot command: "btrfs subvolume snapshot".
.RE
.sp
\-\-send
.RS 4
Allow btrfs send command: "btrfs send".
.RE
.sp
\-\-receive
.RS 4
Allow btrfs receive command: "btrfs receive".
.RE
.sp
\-p, \-\-restrict\-path <path>
.RS 4
Restrict btrfs commands to <path>.
.RE
.sp
\-l, \-\-log
.RS 4
Log ACCEPT and REJECT messages to the system log.
.RE
.sp
\-\-sudo
.RS 4
Allow btrfs commands to be called via sudo. Enable this if you
have "backend btrfs\-progs\-sudo" in your btrbk configuration file.
.RE
.SH "AVAILABILITY"
.sp
Please refer to the btrbk project page \fB\c
.URL "https://digint.ch/btrbk/" "" "\fP"
for further details.
.SH "SEE ALSO"
.sp
\fBbtrbk\fP(1),
\fBbtrbk.conf\fP(5),
\fBbtrfs\fP(8)
.SH "AUTHOR"
.sp
Axel Burri <\c
.MTO "axel\(attty0.ch" "" ">"

ChangeLog
btrbk-0.28.0

  * MIGRATION
    - Replace "ssh_port" options with "ssh://hostname[:port]" notation
      on "volume" or "target" declarations.
    - Check filter statements in your scripts ("btrbk run <filter>"),
      especially when using groups having the same name as subvolumes.
  * Fix table format "resolved" (btrbk list snapshots|backups).
  * Add btrbk-verify: tool for automated backup integrity check based
    on rsync (close #256).
  * Use "ssh://hostname[:port]" notation instead of ambiguous
    "ssh_port" option (now deprecated).
  * Bugfix: correctly handle multiple volume sections with same host,
    but distinct port numbers (virtual machines setups, close #233).
  * Remove selected rows ("*host", "*port") from table output if all
    values are empty.
  * Consider all parent/child relations (close #255).
  * Disallow unrelated parents for btrfs-send if incremental=strict.
  * Add clone sources to btrfs-send if necessary.
  * Add "incremental_resolve" configuration option.
  * Add "incremental_clones" configuration option.
  * Add "noauto" configuration option.
  * Add "--exclude" command line option.
  * Make target_type optional, defaults to "send-receive".
  * Use more sophisticated filter arguments.
  * Allow whitespace separated "group" option.

btrbk-0.27.2

  * Bump dependency >= btrfs-progs-4.12 (as of btrbk-0.27.0).
  * Trigger autofs mount while resolving mountpoints (close #259).
  * Bugfix: fix scheduler when overriding "target_preserve_min" in
    combination with global "target" section.

btrbk-0.27.1

  * MIGRATION
    - Update ssh_filter_btrbk.sh on remote hosts if using
      backend=btrfs-progs-sudo.
  * Bugfix (regression): call "sudo -n readlink" for
    backend=btrfs-progs-sudo (close #257).
  * Bugfix (regression): fix failing "config print" command.
  * Don't print title and blank lines for --format=raw, syslog and
    transaction log.

btrbk-0.27.0

  * MIGRATION
    - Run "btrbk prune --dry-run --print-schedule" and check if any
      snapshots/backups would get deleted [1] [2]. If you want to
      forcibly preserve a snapshot/backup forever, rename it (btrbk
      ignores subvolumes with unknown naming scheme), e.g.:
      "mv mysubvol.YYYYMMDD mysubvol.YYYYMMDD.keep_forever"
    - Update ssh_filter_btrbk.sh on remote hosts (btrbk always calls
      "readlink" and "cat /proc/self/mountinfo").
  * Add "preserve_hour_of_day" configuration option (close #202).
  * Allow backup of filesystem root using "subvolume ." (close #240).
  * Add "-S" command line option (shortcut for --print-schedule).
  * Bugfix: correct scheduling of "first weekly backup in month/year"
    (close #217) [1] [2].
  * Bugfix: add "btrfs filesystem usage" backend mapping (close #222).
  * Bugfix: do not fail (exitcode=10) if targets are not accessible on
    "btrfs snapshot --preserve".
  * Bugfix: if "incremental no" is configured on target, do not keep
    latest common snapshot.
  * Enhance internal data structures:
    - Allow snapshot_dir to be a mountpoint.
    - Search complete target tree for correlated subvolumes.
    - Include snapshots from all mountpoints as candidates (disabled
      due to uptream bug: github.com/kdave/btrfs-progs/issues/96).
    - Read /proc/self/mountinfo instead of /proc/self/mounts.
    - Always read /proc/self/mountinfo.
    - Resolve realpath using readlink(1).
  * Fallback to "asciidoctor" for manpage generation (close #219).

      [1] https://github.com/digint/btrbk/issues/217
      [2] https://github.com/digint/btrbk/commit/719fb5f

btrbk-0.26.1

  * Add "archive_exclude" configuration option.
  * Add warning on redefined configuration option.
  * Bugfix: fix parsing of "openssl_iv_size" configuration option.
  * Bugfix: fix filter statement matching for volume=/ (close #209).

btrbk-0.26.0

  * MIGRATION
    - If you are using raw targets, make sure to run the
      "raw_suffix2sidecar" utility in each target directory.
  * Support for btrfs-progs v4.13.2: adapt parsing of "btrfs sub list"
    output (close #192).
  * Add "resume" command, replacement for "-r, --resume-only" command
    line option (which is now deprecated).
  * Add "snapshot" command (close #150).
  * Add "prune" command.
  * Add "--preserve-snapshots" and "--preserve-backups" options.
  * Add "--wipe" command line option (close #173).
  * Change raw backup format (sidecar file instead of uuid in file).
  * Honor target_preserve for raw targets (delete raw targets).
  * Add symmetric encryption for raw targets (close #157).
  * Add "{snapshot,target,archive}_qgroup_destroy" configuration
    options (close #49, #189).
  * Do not run in "perl taint mode" by default: remove "perl -T" in
    hashbang; hardcode $PATH only if taint mode is enabled.
  * Remove "duration" column from transaction_log/transaction_syslog.
  * Resolve ancestors (recursive on parent_uuid chain) when searching
    for latest common subvolume.
  * Generate man pages from asciidoc (remove raw groff sources).
  * Bugfix: ssh_filter_btrbk: accept mbuffer command (stream_buffer).
  * Bugfix: print correct (end-)time in transaction_log.
  * Bugfix: check path when expanding wildcards (close #181).
  * Bugfix: never show failed deletes in summary (close #183).

btrbk-0.25.1

  * Support for btrfs-progs v4.12: fix parsing of "btrfs sub show"
    output, which now prints relative paths (close #171).
  * Add "stream_buffer" configuration option (close #154).
  * Bugfix: accept "no" for "transaction_log", "transaction_syslog"
    and "lockfile" configuration options.
  * Show "up-to-date" status for backups in "stats" command.
  * Show "correlated" status instead of "orphaned" in "stats" command.
  * Check source subvolumes for readonly and received_uuid flags, and
    abort if one of them is set.

btrbk-0.25.0

  * MIGRATION
    - If you call ssh_filter_btrbk.sh with "--sudo" option, make sure
      to set "backend btrfs-progs-sudo" in btrbk.conf for this host.

ssh_filter_btrbk.sh
#!/bin/bash

set -e
set -u

export PATH=/sbin:/bin:/usr/sbin:/usr/bin

enable_log=
restrict_path_list=
allow_list=
allow_exact_list=
allow_rate_limit=1
allow_stream_buffer=1
allow_compress=1
compress_list="gzip|pigz|bzip2|pbzip2|xz|lzop|lz4"

# note that the backslash is NOT a metacharacter in a POSIX bracket expression!
option_match='-[a-zA-Z0-9=-]+'   # matches short as well as long options
file_match='[0-9a-zA-Z_@+./-]*'  # matches file path (equal to $file_match in btrbk)

log_cmd()
{
    if [[ -n "$enable_log" ]]; then
        logger -p $1 -t ssh_filter_btrbk.sh "$2 (Name: ${LOGNAME:-<unknown>}; Remote: ${SSH_CLIENT:-<unknown>})${3:+: $3}: $SSH_ORIGINAL_COMMAND"
    fi
}

allow_cmd()
{
    allow_list="${allow_list}|$1"
}

allow_exact_cmd()
{
    allow_exact_list="${allow_exact_list}|$1"
}

reject_and_die()
{
    local reason=$1
    log_cmd "auth.err" "btrbk REJECT" "$reason"
    echo "ERROR: ssh_filter_btrbk.sh: ssh command rejected: $reason: $SSH_ORIGINAL_COMMAND" 1>&2
    exit 1
}

run_cmd()
{
    log_cmd "auth.info" "btrbk ACCEPT"
    eval " $SSH_ORIGINAL_COMMAND"
}

reject_filtered_cmd()
{
    if [[ -n "$restrict_path_list" ]]; then
	# match any of restrict_path_list with or without trailing slash,
	# or any file/directory (matching file_match) below restrict_path
	path_match="(${restrict_path_list})(/${file_match})?"
    else
	# match any absolute file/directory (matching file_match)
	path_match="/${file_match}"
    fi

    if [[ -n "$allow_compress" ]]; then
        decompress_match="((${compress_list}) -d -c( -[pT][0-9]+)? \| )?"
        compress_match="( \| (${compress_list}) -c( -[0-9])?( -[pT][0-9]+)?)?"
    else
        decompress_match=
        compress_match=
    fi

    if [[ -n "$allow_stream_buffer" ]]; then
        stream_buffer_match="(mbuffer -q -m [0-9]+[kmg]? \| )?"
    else
        stream_buffer_match=
    fi

    if [[ -n "$allow_rate_limit" ]]; then
        rate_limit_match="( \| pv -q -L [0-9]+[kmgt]?)?"
    else
        rate_limit_match=
    fi

    # allow multiple paths (e.g. "btrfs subvolume snapshot <src> <dst>")
    btrfs_cmd_match="^${decompress_match}${stream_buffer_match}(${allow_list})( ${option_match})*( ${path_match})+${compress_match}${rate_limit_match}$"

    if [[ $SSH_ORIGINAL_COMMAND =~ $btrfs_cmd_match ]] ; then
        return 0
    fi

    exact_cmd_match="^${allow_exact_list}$";
    if [[ $SSH_ORIGINAL_COMMAND =~ $exact_cmd_match ]] ; then
        return 0
    fi

    reject_and_die "disallowed command${restrict_path_list:+ (restrict-path: \"${restrict_path_list//|/\", \"}\")}"
}


# check for "--sudo" option before processing other options
sudo_prefix=
for key; do
    [[ "$key" == "--sudo" ]] && sudo_prefix="sudo -n "
done

while [[ "$#" -ge 1 ]]; do
    key="$1"

    case $key in
      -l|--log)
          enable_log=1
          ;;

      --sudo)
          # already processed above
          ;;

      -p|--restrict-path)
          restrict_path_list="${restrict_path_list}|${2%/}"  # add to list while removing trailing slash
          shift # past argument
          ;;

      -s|--source)
          allow_cmd "${sudo_prefix}btrfs subvolume snapshot"
          allow_cmd "${sudo_prefix}btrfs send"
          ;;

      -t|--target)
          allow_cmd "${sudo_prefix}btrfs receive"
          ;;

      -c|--compress)
          # deprecated option, compression is always allowed
          ;;

      -d|--delete)
          allow_cmd "${sudo_prefix}btrfs subvolume delete"
          ;;

      -i|--info)
          allow_cmd "${sudo_prefix}btrfs subvolume find-new"
          allow_cmd "${sudo_prefix}btrfs filesystem usage"
          ;;

      --snapshot)
          allow_cmd "${sudo_prefix}btrfs subvolume snapshot"
          ;;

      --send)
          allow_cmd "${sudo_prefix}btrfs send"
          ;;

      --receive)
          allow_cmd "${sudo_prefix}btrfs receive"
          ;;

      *)
          echo "ERROR: ssh_filter_btrbk.sh: failed to parse command line option: $key" 1>&2
          exit 1
          ;;
    esac
    shift
done

allow_cmd "${sudo_prefix}btrfs subvolume show"; # subvolume queries are always allowed
allow_exact_cmd "${sudo_prefix}btrfs subvolume list ${file_match}"; # subvolume queries are always allowed
allow_cmd "${sudo_prefix}readlink"              # used to resolve mountpoints
allow_exact_cmd "cat /proc/self/mountinfo"      # used to resolve mountpoints
allow_exact_cmd "cat /proc/self/mounts"         # legacy, for btrbk < 0.27.0

# remove leading "|" on alternation lists
allow_list=${allow_list#\|}
allow_exact_list=${allow_exact_list#\|}
restrict_path_list=${restrict_path_list#\|}

case "$SSH_ORIGINAL_COMMAND" in
    *\.\./*)  reject_and_die "directory traversal"  ;;
    *\$*)     reject_and_die "unsafe character"     ;;
    *\&*)     reject_and_die "unsafe character"     ;;
    *\(*)     reject_and_die "unsafe character"     ;;
    *\{*)     reject_and_die "unsafe character"     ;;
    *\;*)     reject_and_die "unsafe character"     ;;
    *\<*)     reject_and_die "unsafe character"     ;;
    *\>*)     reject_and_die "unsafe character"     ;;
    *\`*)     reject_and_die "unsafe character"     ;;
    *\|*)     [[ -n "$allow_compress" ]] || [[ -n "$allow_rate_limit" ]] || [[ -n "$allow_stream_buffer" ]] || reject_and_die "unsafe character (compression disallowed)" ;;
esac

reject_filtered_cmd
run_cmd
